docs/rfc/rfc-001-storage-engine.rst

```diff
@@ -0,0 +1,179 @@
+===========================================
+RFC 001: Storage Engines and Database Layer
+===========================================
+
+Changelog
+---------
+
+- 2021-04-19: Initial Draft (gist)
+- 2021-09-02: Migrated to RFC folder, with some updates  
+
+Abstract
+--------
+
+The aspect of Tendermint that's responsible for persistence and storage (often
+"the database" internally) represents a bottle neck in the architecture of the
+platform, that the 0.36 release presents a good opportunity to correct. The
+current storage engine layer provides a great deal of flexibility that is
+difficult for users to leverage or benefit from, while also making it harder
+for Tendermint Core developers to deliver improvements on storage engine. This
+RFC discusses the possible improvements to this layer of the system.
+
+Background
+----------
+
+Tendermint has a very thin common wrapper that makes Tendermint itself
+(largely) agnostic to the data storage layer (within the realm of the popular
+key-value/embedded databases.) This flexibility is not particularly useful:
+the benefits of a specific database engine in the context of tendermint is not
+particularly well understood, and the maintenance burden for multiple backends
+is not commensurate with the benefit provided. Additionally, because the data
```

Comment:
I don't entirely get the connection between the interface being very generic and the in-memory test framework. Is the case that "because we have to be able to implement it in-memory, we can't rely on clever storage behaviour because then we'd have to replicate it?" (If so, I agree)

Reply:
In the current implementation, yes, the fact that our tests are all mocked out mean that we can't actually rely on anything. In essence, our current storage engine is a `map[[]byte][]bytes{}` that happens to get written to disk sometimes, and all other storage features have to be implemented in the application.